File: CS/WindowsApplication10/QueryStringBuilder.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraPivotGrid;
using DevExpress.XtraPivotGrid.Data;

namespace WindowsApplication10
{
    class QueryStringBuilder
    {
        static string _select = " select ";
        static string _nonEmpty = " non empty ";
        static string _onColumns = " on columns ";
        static string _onRows = " on rows ";
        static string _from = " from ";
        static string _members = ".members";
        static string _measureSeparator = " , ";
        static string _dimensionSeparator = " * ";
        static string _filedNameSeparator = ".";
        static string _parametersSeparator = ", ";
        static string _generate = "generate";
        static string _curentMember = ".currentmember";
        static string _descendants = "descendants";        

        private static string CurlyBrackets(string str) { return "{" + str + "}"; }
        private static string SquareBrackets(string str) { return "[" + str + "]"; }
        private static string RoundBrackets(string str) { return "(" + str + ")"; }

        
        public static string GetQueryString(PivotGridControl pivot, string fullConnectionString)
        {
            return GetQueryString(pivot, fullConnectionString, true);
        }

        public static string GetQueryString(PivotGridControl pivot, string fullConnectionString, bool nonEmptyBehavior)
        {
            OLAPConnectionStringBuilder builder = new OLAPConnectionStringBuilder(fullConnectionString);
            string cubeName = builder.CubeName; 
            return _select + GetNonEmptyString(nonEmptyBehavior) + GetMeasuresString(pivot) + _onColumns + _parametersSeparator + GetNonEmptyString(nonEmptyBehavior) + GetDimensionsString(pivot) + _onRows + _from + SquareBrackets(cubeName);            
        }

        protected static string GetNonEmptyString(bool nonEmptyBehavior)
        {
            if (nonEmptyBehavior) return _nonEmpty;
            return string.Empty;
        }

        private static string GetDimensionsString(PivotGridControl pivot)
        {
            string dimensionString = string.Empty;
            bool _isFirstItem = true;
            foreach (PivotGridField field in pivot.Fields)
            {
                if (field.Area == PivotArea.DataArea || field.Visible == false || field.InnerGroupIndex > 0) continue;
                if (!_isFirstItem) 
                    dimensionString = dimensionString + _dimensionSeparator;
                else 
                    _isFirstItem = false;
                if (field.InnerGroupIndex < 0)
                    dimensionString = dimensionString + GetDimensionName(field);
                else
                    dimensionString = dimensionString + GetHierarchyString(pivot, field);

            }
            return dimensionString;
        }
        private static string GetMeasuresString(PivotGridControl pivot)
        {
            string measureString = string.Empty;
            bool _isFirstItem = true;
            foreach (PivotGridField field in pivot.GetFieldsByArea(PivotArea.DataArea))
            {
                if (!_isFirstItem)
                    measureString = measureString + _measureSeparator;
                else
                    _isFirstItem = false;
                measureString = measureString + field.FieldName;
            }
            return CurlyBrackets(measureString);
        }

        private static string GetHierarchyString(PivotGridControl pivot, PivotGridField field) 
        {
            return _generate + RoundBrackets(field.FieldName + _members + _parametersSeparator + GetDescendantsString(pivot, field));             
        }
        private static string GetDescendantsString(PivotGridControl pivot, PivotGridField field)         { 
            
            return _descendants + RoundBrackets(GetHierarchyName(field) + _curentMember + _parametersSeparator +  GetLastLevelFieldName(pivot, field));
        }

        private static string GetLastLevelFieldName(PivotGridControl pivot, PivotGridField field)
        {
            PivotGridGroup group = pivot.Groups[field.GroupIndex];
            PivotGridFieldBase lastLevelField =  group[group.Count - 1];
            return lastLevelField.FieldName;
        }

        private static string GetHierarchyName(PivotGridField field) 
        {
            string[] names = field.FieldName.Split(_filedNameSeparator.ToCharArray());
            string name = names[0];
            for (int i = 1; i < names.Length - 1; i++)
                name = name + _filedNameSeparator + names[i];
            return name;
        }

        private static string GetDimensionName(PivotGridField field) { return CurlyBrackets(field.FieldName + _members); }
        private static string GetMeasureName(PivotGridField field) { return field.FieldName; }
    }
}

File: VB/WindowsApplication10/QueryStringBuilder.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraPivotGrid
Imports DevExpress.XtraPivotGrid.Data

Namespace WindowsApplication10
	Friend Class QueryStringBuilder
		Private Shared _select As String = " select "
		Private Shared _nonEmpty As String = " non empty "
		Private Shared _onColumns As String = " on columns "
		Private Shared _onRows As String = " on rows "
		Private Shared _from As String = " from "
		Private Shared _members As String = ".members"
		Private Shared _measureSeparator As String = " , "
		Private Shared _dimensionSeparator As String = " * "
		Private Shared _filedNameSeparator As String = "."
		Private Shared _parametersSeparator As String = ", "
		Private Shared _generate As String = "generate"
		Private Shared _curentMember As String = ".currentmember"
		Private Shared _descendants As String = "descendants"

		Private Shared Function CurlyBrackets(ByVal str As String) As String
			Return "{" & str & "}"
		End Function
		Private Shared Function SquareBrackets(ByVal str As String) As String
			Return "[" & str & "]"
		End Function
		Private Shared Function RoundBrackets(ByVal str As String) As String
			Return "(" & str & ")"
		End Function


		Public Shared Function GetQueryString(ByVal pivot As PivotGridControl, ByVal fullConnectionString As String) As String
			Return GetQueryString(pivot, fullConnectionString, True)
		End Function

		Public Shared Function GetQueryString(ByVal pivot As PivotGridControl, ByVal fullConnectionString As String, ByVal nonEmptyBehavior As Boolean) As String
			Dim builder As New OLAPConnectionStringBuilder(fullConnectionString)
			Dim cubeName As String = builder.CubeName
			Return _select & GetNonEmptyString(nonEmptyBehavior) + GetMeasuresString(pivot) & _onColumns & _parametersSeparator & GetNonEmptyString(nonEmptyBehavior) + GetDimensionsString(pivot) & _onRows & _from & SquareBrackets(cubeName)
		End Function

		Protected Shared Function GetNonEmptyString(ByVal nonEmptyBehavior As Boolean) As String
			If nonEmptyBehavior Then
				Return _nonEmpty
			End If
			Return String.Empty
		End Function

		Private Shared Function GetDimensionsString(ByVal pivot As PivotGridControl) As String
			Dim dimensionString As String = String.Empty
			Dim _isFirstItem As Boolean = True
			For Each field As PivotGridField In pivot.Fields
				If field.Area = PivotArea.DataArea OrElse field.Visible = False OrElse field.InnerGroupIndex > 0 Then
					Continue For
				End If
				If (Not _isFirstItem) Then
					dimensionString = dimensionString & _dimensionSeparator
				Else
					_isFirstItem = False
				End If
				If field.InnerGroupIndex < 0 Then
					dimensionString = dimensionString & GetDimensionName(field)
				Else
					dimensionString = dimensionString & GetHierarchyString(pivot, field)
				End If

			Next field
			Return dimensionString
		End Function
		Private Shared Function GetMeasuresString(ByVal pivot As PivotGridControl) As String
			Dim measureString As String = String.Empty
			Dim _isFirstItem As Boolean = True
			For Each field As PivotGridField In pivot.GetFieldsByArea(PivotArea.DataArea)
				If (Not _isFirstItem) Then
					measureString = measureString & _measureSeparator
				Else
					_isFirstItem = False
				End If
				measureString = measureString & field.FieldName
			Next field
			Return CurlyBrackets(measureString)
		End Function

		Private Shared Function GetHierarchyString(ByVal pivot As PivotGridControl, ByVal field As PivotGridField) As String
			Return _generate & RoundBrackets(field.FieldName & _members & _parametersSeparator & GetDescendantsString(pivot, field))
		End Function
		Private Shared Function GetDescendantsString(ByVal pivot As PivotGridControl, ByVal field As PivotGridField) As String

			Return _descendants & RoundBrackets(GetHierarchyName(field) & _curentMember & _parametersSeparator & GetLastLevelFieldName(pivot, field))
		End Function

		Private Shared Function GetLastLevelFieldName(ByVal pivot As PivotGridControl, ByVal field As PivotGridField) As String
			Dim group As PivotGridGroup = pivot.Groups(field.GroupIndex)
			Dim lastLevelField As PivotGridFieldBase = group(group.Count - 1)
			Return lastLevelField.FieldName
		End Function

		Private Shared Function GetHierarchyName(ByVal field As PivotGridField) As String
			Dim names() As String = field.FieldName.Split(_filedNameSeparator.ToCharArray())
			Dim name As String = names(0)
			For i As Integer = 1 To names.Length - 2
				name = name & _filedNameSeparator & names(i)
			Next i
			Return name
		End Function

		Private Shared Function GetDimensionName(ByVal field As PivotGridField) As String
			Return CurlyBrackets(field.FieldName & _members)
		End Function
		Private Shared Function GetMeasureName(ByVal field As PivotGridField) As String
			Return field.FieldName
		End Function
	End Class
End Namespace
